From 12a031b3a91144261624e033f68fcce5a609d18a Mon Sep 17 00:00:00 2001
From: Lars Uebernickel <lars.uebernickel@canonical.com>
Date: Wed, 8 May 2013 13:21:55 -0700
Subject: [PATCH] Allow custom menu items from gtk_menu_new_from_model

Provide GtkMenuItemFactory, an interface that can be implemented by
extension points. It has one function: create a GtkMenuItem for a
GMenuItem and a GActionGroup, keyed by the type given in the
x-canonical-type attribute.

This is not considered to be public API.
---
 gtk/Makefile.am          |   2 +
 gtk/gtk.h                |   1 +
 gtk/gtkmenuitemfactory.c |  38 ++++++++++++
 gtk/gtkmenuitemfactory.h |  57 ++++++++++++++++++
 gtk/gtkmodelmenu.c       | 147 ++++++++++++++++++++++++++++++++++++++++++++++-
 5 files changed, 244 insertions(+), 1 deletion(-)
 create mode 100644 gtk/gtkmenuitemfactory.c
 create mode 100644 gtk/gtkmenuitemfactory.h

diff --git a/gtk/Makefile.am b/gtk/Makefile.am
index e164c90..717f8bf 100644
--- a/gtk/Makefile.am
+++ b/gtk/Makefile.am
@@ -277,6 +277,7 @@ gtk_public_h_sources = 		\
 	gtklockbutton.h		\
 	gtkmain.h		\
 	gtkmenu.h		\
+	gtkmenuitemfactory.h	\
 	gtkmenubar.h		\
 	gtkmenubutton.h		\
 	gtkmenuitem.h		\
@@ -765,6 +766,7 @@ gtk_base_c_sources = 		\
 	gtkmisc.c		\
 	gtkmnemonichash.c	\
 	gtkmodelmenu.c		\
+	gtkmenuitemfactory.c	\
 	gtkmodelmenuitem.c	\
 	gtkmodifierstyle.c	\
 	gtkmodules.c		\
diff --git a/gtk/gtk.h b/gtk/gtk.h
index e95494f..c384979 100644
--- a/gtk/gtk.h
+++ b/gtk/gtk.h
@@ -223,6 +223,7 @@
 #include <gtk/gtkwidget.h>
 #include <gtk/gtkwidgetpath.h>
 #include <gtk/gtkwindow.h>
+#include <gtk/gtkmenuitemfactory.h>
 
 #ifndef GTK_DISABLE_DEPRECATED
 #include <gtk/deprecated/gtkcolorsel.h>
diff --git a/gtk/gtkmenuitemfactory.c b/gtk/gtkmenuitemfactory.c
new file mode 100644
index 0000000..007699a
--- /dev/null
+++ b/gtk/gtkmenuitemfactory.c
@@ -0,0 +1,38 @@
+/*
+* Copyright 2013 Canonical Ltd.
+*
+* This program is free software: you can redistribute it and/or modify it
+* under the terms of the GNU General Public License version 3, as published
+* by the Free Software Foundation.
+*
+* This program is distributed in the hope that it will be useful, but
+* WITHOUT ANY WARRANTY; without even the implied warranties of
+* MERCHANTABILITY, SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR
+* PURPOSE.  See the GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License along
+* with this program.  If not, see <http://www.gnu.org/licenses/>.
+*
+* Authors:
+*     Lars Uebernickel <lars.uebernickel@canonical.com>
+*/
+
+#include "gtkmenuitemfactory.h"
+
+G_DEFINE_INTERFACE_WITH_CODE (GtkMenuItemFactory, gtk_menu_item_factory, G_TYPE_OBJECT,
+  GIOExtensionPoint *ep = g_io_extension_point_register (GTK_MENU_ITEM_FACTORY_EXTENSION_POINT_NAME);
+  g_io_extension_point_set_required_type (ep, g_define_type_id);)
+
+static void
+gtk_menu_item_factory_default_init (GtkMenuItemFactoryInterface *iface)
+{
+}
+
+GtkMenuItem *
+gtk_menu_item_factory_create_menu_item (GtkMenuItemFactory *factory,
+                                        const gchar        *type,
+                                        GMenuItem          *menuitem,
+                                        GActionGroup       *actions)
+{
+  return GTK_MENU_ITEM_FACTORY_GET_IFACE (factory)->create_menu_item (factory, type, menuitem, actions);
+}
diff --git a/gtk/gtkmenuitemfactory.h b/gtk/gtkmenuitemfactory.h
new file mode 100644
index 0000000..ffeb8da
--- /dev/null
+++ b/gtk/gtkmenuitemfactory.h
@@ -0,0 +1,57 @@
+/*
+* Copyright 2013 Canonical Ltd.
+*
+* This program is free software: you can redistribute it and/or modify it
+* under the terms of the GNU General Public License version 3, as published
+* by the Free Software Foundation.
+*
+* This program is distributed in the hope that it will be useful, but
+* WITHOUT ANY WARRANTY; without even the implied warranties of
+* MERCHANTABILITY, SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR
+* PURPOSE.  See the GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License along
+* with this program.  If not, see <http://www.gnu.org/licenses/>.
+*
+* Authors:
+*     Lars Uebernickel <lars.uebernickel@canonical.com>
+*/
+
+#ifndef __GTK_MENU_ITEM_FACTORY_H__
+#define __GTK_MENU_ITEM_FACTORY_H__
+
+#include <gio/gio.h>
+#include "gtkmenuitem.h"
+
+G_BEGIN_DECLS
+
+#define GTK_TYPE_MENU_ITEM_FACTORY         (gtk_menu_item_factory_get_type ())
+#define GTK_MENU_ITEM_FACTORY(o)           (G_TYPE_CHECK_INSTANCE_CAST ((o), GTK_TYPE_MENU_ITEM_FACTORY, GtkMenuItemFactory))
+#define GTK_IS_MENU_ITEM_FACTORY(o)        (G_TYPE_CHECK_INSTANCE_TYPE ((o), GTK_TYPE_MENU_ITEM_FACTORY))
+#define GTK_MENU_ITEM_FACTORY_GET_IFACE(o) (G_TYPE_INSTANCE_GET_INTERFACE ((o), GTK_TYPE_MENU_ITEM_FACTORY, GtkMenuItemFactoryInterface))
+
+#define GTK_MENU_ITEM_FACTORY_EXTENSION_POINT_NAME "gtk-menu-item-factory"
+
+typedef struct _GtkMenuItemFactoryInterface GtkMenuItemFactoryInterface;
+typedef struct _GtkMenuItemFactory          GtkMenuItemFactory;
+
+struct _GtkMenuItemFactoryInterface
+{
+  GTypeInterface iface;
+
+  GtkMenuItem * (*create_menu_item)  (GtkMenuItemFactory *factory,
+                                      const gchar        *type,
+                                      GMenuItem          *menuitem,
+                                      GActionGroup       *actions);
+};
+
+GType                   gtk_menu_item_factory_get_type      (void);
+
+GtkMenuItem *           gtk_menu_item_factory_create_menu_item (GtkMenuItemFactory *factory,
+                                                                const gchar        *type,
+                                                                GMenuItem          *menuitem,
+                                                                GActionGroup       *actions);
+
+G_END_DECLS
+
+#endif
diff --git a/gtk/gtkmodelmenu.c b/gtk/gtkmodelmenu.c
index 0c3cc5e..1f4c049 100644
--- a/gtk/gtkmodelmenu.c
+++ b/gtk/gtkmodelmenu.c
@@ -28,6 +28,8 @@
 #include "gtkseparatormenuitem.h"
 #include "gtkmodelmenuitem.h"
 #include "gtkapplicationprivate.h"
+#include "gtkwidgetprivate.h"
+#include "gtkmenuitemfactory.h"
 
 #define MODEL_MENU_WIDGET_DATA "gtk-model-menu-widget-data"
 
@@ -72,6 +74,146 @@ gtk_model_menu_binding_free (gpointer data)
   g_slice_free (GtkModelMenuBinding, binding);
 }
 
+static gboolean
+object_class_has_property (GObjectClass *class,
+                           const gchar  *name,
+                           GType         type)
+{
+  GParamSpec *pspec;
+
+  pspec = g_object_class_find_property (class, name);
+  return pspec != NULL && G_PARAM_SPEC_VALUE_TYPE (pspec) == type;
+}
+
+static GtkMenuItem *
+gtk_model_menu_create_item_from_type (GType      type,
+                                      GMenuItem *menuitem)
+{
+  GObjectClass *class;
+  GtkMenuItem *item = NULL;
+
+  class = g_type_class_ref (type);
+
+  if (g_type_is_a (type, GTK_TYPE_MENU_ITEM) &&
+      object_class_has_property (class, "menu-item", G_TYPE_MENU_ITEM) &&
+      object_class_has_property (class, "action-group", G_TYPE_ACTION_GROUP))
+    {
+      item = g_object_new (type,
+                           "menu-item", menuitem,
+                           NULL);
+
+      g_object_set (item,
+                    "action-group", _gtk_widget_get_action_muxer (GTK_WIDGET (item)),
+                    NULL);
+    }
+  else
+    {
+      g_warning ("gtk_menu_new_from_model: cannot create menu item from type '%s'.\n"
+                 "  Does it derive from GtkMenuItem and have 'menu-item' and 'action-group' properties?",
+                 g_type_name (type));
+    }
+
+  g_type_class_unref (class);
+  return item;
+}
+
+static GtkMenuItem *
+gtk_model_menu_create_custom_item_from_factory (const gchar  *typename,
+                                                GMenuItem    *menuitem,
+                                                GActionGroup *actions)
+{
+  GList *it;
+  static GList *factories = NULL;
+
+  if (factories == NULL)
+    {
+      GIOExtensionPoint *ep;
+
+      g_type_ensure (GTK_TYPE_MENU_ITEM_FACTORY);
+      ep = g_io_extension_point_lookup (GTK_MENU_ITEM_FACTORY_EXTENSION_POINT_NAME);
+      for (it = g_io_extension_point_get_extensions (ep); it != NULL; it = it->next)
+        {
+          GIOExtension *ext = it->data;
+          GtkMenuItemFactory *factory;
+
+          factory = g_object_new (g_io_extension_get_type (ext), NULL);
+          factories = g_list_prepend (factories, factory);
+        }
+      factories = g_list_reverse (factories);
+    }
+
+  for (it = factories; it != NULL; it = it->next)
+    {
+      GtkMenuItemFactory *factory = it->data;
+      GtkMenuItem *item;
+
+      item = gtk_menu_item_factory_create_menu_item (factory, typename, menuitem, actions);
+      if (item)
+        return item;
+    }
+
+  return NULL;
+}
+
+static GtkMenuItem *
+gtk_model_menu_create_custom_item (GMenuModel  *menu,
+                                   gint         item_index,
+                                   const gchar *action_namespace,
+                                   GtkWidget   *parent)
+{
+  GMenuItem *menuitem;
+  gchar *typename;
+  GtkMenuItem *item = NULL;
+
+  menuitem = g_menu_item_new_from_model (menu, item_index);
+  if (action_namespace)
+    {
+      gchar *action;
+      gchar *fullname;
+
+      g_menu_item_get_attribute (menuitem, G_MENU_ATTRIBUTE_ACTION, "s", &action);
+      fullname = g_strconcat (action_namespace, ".", action, NULL);
+      g_menu_item_set_attribute (menuitem, G_MENU_ATTRIBUTE_ACTION, "s", fullname);
+
+      g_free (action);
+      g_free (fullname);
+    }
+
+  if (g_menu_model_get_item_attribute (menu, item_index, "x-canonical-type", "s", &typename))
+    {
+      GActionGroup *actions;
+
+      /* Passing the parent muxer is wrong, but we'll only have access
+       * to the menuitem's muxer after the widget has been created.
+       * Thus we'd need some other form of passing the action group to
+       * the widget, which would complicate things for no practical
+       * reason: the panel service is the only consumer of this API and
+       * it will never call gtk_widget_insert_action_group() on the
+       * returned menu item.
+       */
+      actions = G_ACTION_GROUP (_gtk_widget_get_action_muxer (parent));
+      item = gtk_model_menu_create_custom_item_from_factory (typename, menuitem, actions);
+
+      /* continue supporting GTypes for now (the "old" behavior) */
+      if (item == NULL)
+        {
+          GType type;
+
+          type = g_type_from_name (typename);
+          if (type > 0)
+            item = gtk_model_menu_create_item_from_type (type, menuitem);
+        }
+
+      if (item == NULL)
+        g_warning ("gtk_menu_new_from_model: cannot find type '%s'", typename);
+
+      g_free (typename);
+    }
+
+  g_object_unref (menuitem);
+  return item;
+}
+
 static void
 gtk_model_menu_binding_append_item (GtkModelMenuBinding  *binding,
                                     GMenuModel           *model,
@@ -106,7 +248,10 @@ gtk_model_menu_binding_append_item (GtkModelMenuBinding  *binding,
     {
       GtkMenuItem *item;
 
-      item = gtk_model_menu_item_new (model, item_index, action_namespace);
+      item = gtk_model_menu_create_custom_item (model, item_index, action_namespace, GTK_WIDGET (binding->shell));
+      if (item == NULL)
+        item = gtk_model_menu_item_new (model, item_index, action_namespace);
+
       gtk_menu_shell_append (binding->shell, GTK_WIDGET (item));
       gtk_widget_show (GTK_WIDGET (item));
       binding->n_items++;
-- 
1.8.1.2

